src/Microsoft.AspNetCore.Http/Internal/ServerAddressTemplate.cs

```diff
@@ -0,0 +1,151 @@
+
+            var isUnixPipe = url.IndexOf(Constants.UnixPipeHostPrefix, schemeDelimiterEnd, StringComparison.Ordinal) == schemeDelimiterEnd;
+
+            int pathDelimiterStart;
+            int pathDelimiterEnd;
+            if (!isUnixPipe)
+            {
+                pathDelimiterStart = url.IndexOf("/", schemeDelimiterEnd, StringComparison.Ordinal);
+                pathDelimiterEnd = pathDelimiterStart;
+            }
+            else
+            {
+                pathDelimiterStart = url.IndexOf(":", schemeDelimiterEnd + Constants.UnixPipeHostPrefix.Length, StringComparison.Ordinal);
+                pathDelimiterEnd = pathDelimiterStart + ":".Length;
+            }
+
+            if (pathDelimiterStart < 0)
+            {
+                pathDelimiterStart = pathDelimiterEnd = url.Length;
+            }
+
+            var serverAddress = new ServerAddressTemplate();
+            serverAddress.Scheme = url.Substring(0, schemeDelimiterStart);
+
+            var hasSpecifiedPort = false;
+            if (!isUnixPipe)
+            {
+                int portDelimiterStart = url.LastIndexOf(":", pathDelimiterStart - 1, pathDelimiterStart - schemeDelimiterEnd, StringComparison.Ordinal);
+                if (portDelimiterStart >= 0)
+                {
+                    int portDelimiterEnd = portDelimiterStart + ":".Length;
+
+                    string portString = url.Substring(portDelimiterEnd, pathDelimiterStart - portDelimiterEnd);
+                    int portNumber;
+                    if (int.TryParse(portString, NumberStyles.Integer, CultureInfo.InvariantCulture, out portNumber))
+                    {
+                        hasSpecifiedPort = true;
+                        serverAddress.Host = url.Substring(schemeDelimiterEnd, portDelimiterStart - schemeDelimiterEnd);
+                        serverAddress.Port = portNumber;
+                    }
+                }
+
+                if (!hasSpecifiedPort)
+                {
+                    if (string.Equals(serverAddress.Scheme, "http", StringComparison.OrdinalIgnoreCase))
+                    {
+                        serverAddress.Port = 80;
+                    }
+                    else if (string.Equals(serverAddress.Scheme, "https", StringComparison.OrdinalIgnoreCase))
+                    {
+                        serverAddress.Port = 443;
+                    }
+                }
+            }
+
+            if (!hasSpecifiedPort)
+            {
+                serverAddress.Host = url.Substring(schemeDelimiterEnd, pathDelimiterStart - schemeDelimiterEnd);
+            }
+
+            if (string.IsNullOrEmpty(serverAddress.Host))
+            {
+                throw new FormatException($"Invalid url: '{url}'");
+            }
+
+            if (url[url.Length - 1] == '/')
+            {
+                serverAddress.PathBase = url.Substring(pathDelimiterEnd, url.Length - pathDelimiterEnd - 1);
```

Comment:
PathBase only exists for legacy reasons. I think it's better to just throw from this method if a path is present.

The only place PathBase is used in Kestrel is [here](https://github.com/aspnet/KestrelHttpServer/blob/668f8e3b4b4f6e74e7a36433439847575889058e/src/Kestrel.Core/Internal/AddressBinder.cs#L129-L132) just to verify it's null or empty.

Reply:
HttpSys is going to start using this too, and it requires PathBase.